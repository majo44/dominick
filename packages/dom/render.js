import {VNODE, executeInHandleContext} from '@dominick/core/index.js';
import {handleHookShouldUpdate} from './hooks/shouldUpdate.js';
import {handleHookUseEffectPostUpdate} from './hooks/useEffect.js';
import {handleHookUseState} from './hooks/useState.js';
import {parseDynamicAttributes} from './attrs.js';
import {applyEvents} from './event.js';
import {getLocalContext, clearLocalContext, setLocalContext} from './context/localContext.js';
import {clearGlobalContext, setGlobalContext, getGlobalContext} from './context/globalContext.js';
import {hookHandle} from './hooks/hookHandle.js';

/* global IncrementalDOM */

/**
 * @typedef {import('../core/h.js').VNode} VNode
 * @typedef {Element & {[LOCAL_CONTEXT]: *}} AugmentedElement
 * @typedef {VNode | (function & {[VNODE]: boolean})|string|number|undefined|null} PrimitiveVNodeChild
 * @typedef {PrimitiveVNodeChild|Array.<PrimitiveVNodeChild>} VNodeChild
 */

const LOCAL_CONTEXT = Symbol();
const MARKER = '_@_@_@_';

const {
    elementOpen, text, elementClose, patch, skipNode, attributes,
    applyProp, applyAttr, currentPointer, elementVoid
} = IncrementalDOM;

/**
 * @param {Element} el
 * @param {string} name
 * @param {string} value
 */
attributes.value = (el, name, value) => {
    applyProp(el, name, value);
    applyAttr(el, name, value);
};

/**
 * @param {Element} el
 * @param {string} name
 * @param {string} value
 */
attributes.disabled = attributes.checked = (el, name, value) => {
    // for boolean attrs idom passes values as '' for true and undefined for false
    const newValue = value === '';
    applyProp(el, name, newValue);
    applyAttr(el, name, value);
};

/**
 * @type {number}
 */
let paramsPointer = 0;

/**
 * @param {Node} node
 */
function onStart(node) {
    paramsPointer = 0;
    return onChildren(node);
}

/**
 * @param {Node} node
 * @return {(function(Array.<*>): *)}
 */
function onChildren(node) {
    if (node.childNodes.length > 0) {
        const children = Array.prototype
            .map.call(node.childNodes, onElement)
            .filter(/** @param {*} i */ (i) => !!i);
        const count = children.length;
        if (count === 1) {
            const child = children[0];
            return (params) => [child(params)];
        } else {
            return function (params) {
                let res = [];
                for (let i = 0; i < count; i++) {
                    res[i] = children[i](params);
                }
                return res;
            };
        }
    } else {
        return undefined;
    }
}

/**
 * @param {*} child
 * @return {*}
 */
function renderChild(child) {
    switch (typeof child) {
        case 'function': {
            if (!child[VNODE]) {
                try {
                    const myPointer = /** @type {AugmentedElement} */ (currentPointer());
                    let skip = false;
                    let currentLocalContext = myPointer ? myPointer[LOCAL_CONTEXT] : undefined;
                    setLocalContext(currentLocalContext);
                    const nextChild = /** @type {VNodeChild} */ (executeInHandleContext(child, hookHandle));
                    currentLocalContext = getLocalContext();
                    if (currentLocalContext) {
                        skip = !handleHookShouldUpdate(currentLocalContext);
                    }
                    if (!skip) {
                        const [dom] = /** @type {Array.<AugmentedElement>} */ (renderChild(nextChild));
                        dom[LOCAL_CONTEXT] = currentLocalContext;
                        handleHookUseEffectPostUpdate(currentLocalContext);
                        handleHookUseState(currentLocalContext, () => rerender(dom, child, getGlobalContext()));
                        return dom;
                    } else {
                        skipNode();
                        return myPointer;
                    }
                } finally {
                    clearLocalContext();
                }
            } else {
                return child({createVNode, hookHandle});
            }
        }
        case 'string':
            return text(child);
        case 'number':
            return text(child.toString());
        case 'boolean':
        case 'undefined':
            return;
        default: {
            if (Array.isArray(child)) {
                const count = child.length;
                const res = [];
                for (let i = 0; i < count; i++) {
                    res[i] = renderChild(child[i]);
                }
                return res;
            } else {
                return;
            }
        }
    }
}

/**
 * @param {Element} element
 */
function onElement(element) {
    if (element.nodeType === 3 && element.nodeValue) {
        const parts = element.nodeValue.split(MARKER);
        if (parts.length < 2) {
            return () => {
                return text(parts[0]);
            };
        } else {
            const pointer = paramsPointer;
            paramsPointer = paramsPointer + parts.length - 1;
            return (
                /**
                 * @param {Array.<*>} params
                 */
                (params) => {
                    let localPointer = pointer;
                    let result = [];
                    for (let i = 0; i < parts.length; i++) {
                        if (parts[i]) {
                            result.push(text(parts[i]));
                        }
                        if (i < parts.length - 1) {
                            result.push(renderChild(params[localPointer++]));
                        }
                    }
                    return result;
                });
        }
    }

    /**
     * @type {string}
     */
    let tag = element.tagName;
    /**
     * @type {Array.<string>}
     */
    let staticAttrs = [];
    /**
     * @type {Array.<{name: string, pointer: number}>}
     */
    let dynamicAttrs = [];

    Array.prototype.forEach.call(element.attributes,
        /**
         * @param {Attr} attr
         */
        (attr) => {
            if (attr.value === MARKER) {
                dynamicAttrs.push(({name: attr.name, pointer: paramsPointer++}));
            } else {
                staticAttrs.push(attr.name);
                staticAttrs.push(attr.value);
            }
        });

    const children = onChildren(element);

    return (
        /**
         * @param {Array.<*>} params
         */
        (params) => {
            const {key, ref, attrs, events} = parseDynamicAttributes(dynamicAttrs, params);
            /**
             * @type {AugmentedElement}
             */
            let dom;
            if (children) {
                dom = /** @type {AugmentedElement} **/ (elementOpen(tag, key, undefined, ...staticAttrs, ...attrs));
                children(params);
                elementClose(tag);
            } else {
                dom = /** @type {AugmentedElement} **/ (elementVoid(tag, key, staticAttrs, ...attrs));
            }
            events && applyEvents(dom, events);
            ref && ref(dom);
            return dom;
        });
}

/**
 * @param {Array.<string>} literal
 */
function createVNode(literal) {
    const raw = literal.join(MARKER);
    const parser = new DOMParser();
    const parsed = parser.parseFromString(raw, 'application/xml');
    return onStart(parsed);
}

/**
 * @param {Element} previousDom
 * @param {function} vnode
 * @param {*} context
 */
function rerender(previousDom, vnode, context) {
    patch(previousDom.parentElement, () => {
        const len = previousDom.parentNode.childNodes.length;
        for (let i = 0; i < len; i++) {
            if (currentPointer() !== previousDom) {
                skipNode();
            } else {
                setGlobalContext(context);
                renderChild(vnode);
                clearGlobalContext();
            }
        }
    }, {createVNode});
}


/**
 * Renders markup under provided parent node.
 * @param {Element} where the target of rendering
 * @param {VNode | function():VNode } what markup which have to be rendered
 * @param {*?} globalContext the custom global context which is available
 *      within the functional component by use getContext hook
 */
export function render(where, what, globalContext) {
    if (typeof what === 'function') {
        setGlobalContext(globalContext);
        patch(where, () => renderChild(what), {createVNode});
        clearGlobalContext();
    } else {
        throw new Error('Illegal argument exception.');
    }
}