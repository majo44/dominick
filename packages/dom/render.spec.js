import { h, shouldUpdate, useContext, useEffect, useRef, useState, VNODE  } from '../core/index.js';
import { render } from './render.js';

const { notifications } = IncrementalDOM;

describe('render', () => {
    afterEach(() => window.document.body.innerHTML = '');

    it('should throw exception if not function is provided as item to render', () => {
        expect(() => render(window.document.body, 1)).toThrow();
    });
    it('should render nothing if try to render object', () => {
        render(window.document.body, h`<div>${({})}</div>`);
        expect(window.document.body.innerHTML).toMatchSnapshot();
    });
    it('is renders simple markup', () => {
        const app = (name) => h`<div>Hello ${name}</div>`;
        render(window.document.body, app('Majo'));
        expect(window.document.body.innerHTML).toMatchSnapshot();
    });
    it('should render number', () => {
        render(window.document.body, h`<div>${1}</div>`);
        expect(window.document.body.innerHTML).toMatchSnapshot();
    });
    it('should not render boolean', () => {
        render(window.document.body, h`<div>${true}${false}</div>`);
        expect(window.document.body.innerHTML).toMatchSnapshot();
    });
    it('should render functional component', () => {
        const component = (name) => () => h`<span>${name}</span>`;
        render(window.document.body, h`<div>${component('Majo')}</div>`);
        expect(window.document.body.innerHTML).toMatchSnapshot();
    });
    it('should render array of children', () => {
        render(window.document.body, h`<div>${[h`<span>Hello</span>`, h`<span>Pawel</span>`]}</div>`);
        expect(window.document.body.innerHTML).toMatchSnapshot();
    });
    it('should render empty tag', () => {
        render(window.document.body, h`<div><span></span></div>`);
        expect(window.document.body.innerHTML).toMatchSnapshot();
    });
    it('should properly set value attribute also as property', () => {
        render(window.document.body, h`<input value="${'value'}"/>`);
        expect(window.document.body.firstElementChild.value).toMatchSnapshot();
        expect(window.document.body.innerHTML).toMatchSnapshot();
    });
    it('should properly set checked true attribute also as property', () => {
        render(window.document.body, h`<input checked="${true}"/>`);
        expect(window.document.body.firstElementChild.checked).toMatchSnapshot();
        expect(window.document.body.innerHTML).toMatchSnapshot();
    });
    it('should properly set checked false attribute also as property', () => {
        render(window.document.body, h`<input checked="${false}"/>`);
        expect(window.document.body.firstElementChild.checked).toMatchSnapshot();
        expect(window.document.body.innerHTML).toMatchSnapshot();
    });
    it('should properly set disabled true attribute also as property', () => {
        render(window.document.body, h`<input disabled="${true}"/>`);
        expect(window.document.body.firstElementChild.disabled).toMatchSnapshot();
        expect(window.document.body.innerHTML).toMatchSnapshot();
    });
    it('should properly set disabled false attribute also as property', () => {
        render(window.document.body, h`<input disabled="${false}"/>`);
        expect(window.document.body.firstElementChild.disabled).toMatchSnapshot();
        expect(window.document.body.innerHTML).toMatchSnapshot();
    });
    it('should support reuse of dom', () => {
        let count = 0;
        notifications.nodesCreated = (nodes) => count = count + nodes.length;
        window.document.body.innerHTML = '<div>Hello<span>Majo</span></div>';
        render(window.document.body, () => h`<div>Hello<span>Majo</span></div>`);
        expect(count).toBe(0);
    });
    it('should support shouldUpdate and not rerender if value is not changed', () => {
        const markup = jest.fn(h`<span>Majo</span>`);
        markup[VNODE] = true;
        const component = () => () => {
            shouldUpdate(1);
            return markup;
        };
        render(window.document.body, component());
        render(window.document.body, component());
        expect(markup).toBeCalledTimes(1);
    });
    it('should support rerender of child component', () => {
        let setNameGlobal;
        const component = () => () => {
            const [name, setName] = useState('Majo');
            setNameGlobal = setName;
            return h`<span>${name}</span>`;
        };
        render(window.document.body, () => h`<div>
            <span>Header</span>
            ${component()}
            <span>Footer</span>
        </div>`);
        expect(window.document.body.innerHTML).toMatchSnapshot();
        setNameGlobal('Olo');
        expect(window.document.body.innerHTML).toMatchSnapshot();
    });
    it('should support dynamic and static attributes', () => {
        const component = (classe) => () => {
            return h`<span class="${classe}" style="color: blue;">Name</span>`;
        };
        render(window.document.body, component('aclasse'));
        expect(window.document.body.innerHTML).toMatchSnapshot();
    });
});