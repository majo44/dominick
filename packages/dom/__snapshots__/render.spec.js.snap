// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render is renders simple markup 1`] = `"<div>Hello Majo</div>"`;

exports[`render should not render boolean 1`] = `"<div></div>"`;

exports[`render should properly set checked false attribute also as property 1`] = `false`;

exports[`render should properly set checked false attribute also as property 2`] = `"<input>"`;

exports[`render should properly set checked true attribute also as property 1`] = `true`;

exports[`render should properly set checked true attribute also as property 2`] = `"<input checked=\\"\\">"`;

exports[`render should properly set disabled false attribute also as property 1`] = `false`;

exports[`render should properly set disabled false attribute also as property 2`] = `"<input>"`;

exports[`render should properly set disabled true attribute also as property 1`] = `true`;

exports[`render should properly set disabled true attribute also as property 2`] = `"<input disabled=\\"\\">"`;

exports[`render should properly set value attribute also as property 1`] = `"value"`;

exports[`render should properly set value attribute also as property 2`] = `"<input value=\\"value\\">"`;

exports[`render should render array of children 1`] = `"<div><span>Hello</span><span>Pawel</span></div>"`;

exports[`render should render empty tag 1`] = `"<div><span></span></div>"`;

exports[`render should render functional component 1`] = `"<div><span>Majo</span></div>"`;

exports[`render should render nothing if try to render object 1`] = `"<div></div>"`;

exports[`render should render number 1`] = `"<div>1</div>"`;

exports[`render should support dynamic and static attributes 1`] = `"<span style=\\"color: blue;\\" class=\\"aclasse\\">Name</span>"`;

exports[`render should support rerender of child component 1`] = `
"<div>
            <span>Header</span>
            <span>Majo</span>
            <span>Footer</span>
        </div>"
`;

exports[`render should support rerender of child component 2`] = `
"<div>
            <span>Header</span>
            <span>Olo</span>
            <span>Footer</span>
        </div>"
`;
